File: loggalitic/src/main/java/com/android/oleksandrpriadko/loggalitic/analytics/policy/Policy.java
package com.android.oleksandrpriadko.loggalitic.analytics.policy;

import com.android.oleksandrpriadko.loggalitic.analytics.AnalyticsEvent;

/**
 * Created by Oleksandr Priadko.
 * 7/9/17
 * <p>
 * Determines levels of reporting analytics events and logs. Implement yours or
 * use {@link DefaultPolicy}
 */
public abstract class Policy {

    private int mLogLevel;
    private int mLogToEventLevel;
    private boolean mIsDebug;

    /**
     * @param logLevel        max log level which can be reported. e.g. if mLogLevel is DEBUG -
     *                        ERROR will not be placed
     * @param logToEventLevel max log level which can de sent to analytics service.
     * @param debug           are we in debug mode?
     */
    public Policy(int logLevel, int logToEventLevel, boolean debug) {
        this.mLogLevel = logLevel;
        this.mLogToEventLevel = logToEventLevel;
        this.mIsDebug = debug;
    }

    protected final int getLogLevel() {
        return this.mLogLevel;
    }

    protected final int getLogToEventLevel() {
        return this.mLogToEventLevel;
    }

    public abstract boolean isEventAllowed(AnalyticsEvent event);

    public abstract boolean isLogAllowed(String tag, int level, boolean force);

    public abstract boolean isLogReportAllowed(String tag, int level, boolean force);

    public abstract boolean isLogAllowed(String tag, int level);

    public final boolean isDebug() {
        return mIsDebug;
    }

    protected final String getTag() {
        return this.getClass().getSimpleName();
    }
}

File: core/src/main/java/com/android/oleksandrpriadko/core/policy/Policy.kt
package com.android.oleksandrpriadko.core.policy

import androidx.annotation.IntDef

/**
 * Determines levels of reporting analytics events and logs. Implement yours or use one
 * from loggalitic module
 */
abstract class Policy
/**
 * @param logLevel        max log level which can be reported. e.g. if mLogLevel is DEBUG -
 * ERROR will not be placed
 * @param logToEventLevel max log level which can de sent to analytics service.
 * @param isDebug           are we in debug mode?
 */
(@param:LogLevel @get:LogLevel
 protected val logLevel: Int,
 @param:LogLevel @get:LogLevel
 protected val logToEventLevel: Int,
 val isDebug: Boolean) {

    protected val tag: String
        get() = this.javaClass.simpleName

    abstract fun isEventAllowed(event: AnalyticsEvent): Boolean

    abstract fun isLogAllowed(tag: String, level: Int, force: Boolean): Boolean

    abstract fun isLogReportAllowed(tag: String, level: Int, force: Boolean): Boolean

    abstract fun isLogAllowed(tag: String, level: Int): Boolean

    @IntDef(VERBOSE, DEBUG, INFO, WARN, ERROR, ASSERT)
    annotation class LogLevel

    companion object {

        const val VERBOSE = 2
        const val DEBUG = 3
        const val INFO = 4
        const val WARN = 5
        const val ERROR = 6
        const val ASSERT = 7

        val NOT_SET = object : Policy(
                logLevel = ASSERT,
                logToEventLevel = ASSERT,
                isDebug = false) {
            override fun isEventAllowed(event: AnalyticsEvent): Boolean = false

            override fun isLogAllowed(tag: String, level: Int, force: Boolean): Boolean = false

            override fun isLogReportAllowed(tag: String, level: Int, force: Boolean): Boolean = false

            override fun isLogAllowed(tag: String, level: Int): Boolean = false

        }
    }
}
